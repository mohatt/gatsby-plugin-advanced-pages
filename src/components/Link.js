import React from 'react'
import { Link as GatsbyLink } from 'gatsby'
import { string, object } from 'prop-types'
import { routeExists, generatePath } from '../api'

const Link = ({ to, params, scope, ...props }) => {
  if (to && routeExists(to)) {
    try {
      to = generatePath(to, params, scope, true)
    } catch (e) {
      console.error(
        'Warning: Invalid route params supplied to `Link` component: ' + e.message
      )
    }
  }

  return (
    <GatsbyLink to={to} {...props} />
  )
}

Link.propTypes = {
  to: string.isRequired,
  params: object,
  scope: string
}

export default Link
