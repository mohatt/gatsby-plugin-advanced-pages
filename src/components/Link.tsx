import React, { ReactNode } from 'react'
import { Link as GatsbyLink, GatsbyLinkProps } from 'gatsby'
import { string, object } from 'prop-types'
import { routeExists, generatePath } from '../api'
import type { RouteScope } from '../gatsby/types'

export interface LinkProps<TState> extends GatsbyLinkProps<TState> {
  to: string
  params?: Record<string, any>
  scope?: RouteScope
}

export const Link = <TState,>({ to, params, scope, ...props }: LinkProps<TState>): ReactNode => {
  if (to != null && routeExists(to)) {
    try {
      to = generatePath(to, params, scope, true)
    } catch (e) {
      if (process.env.NODE_ENV !== 'production') {
        console.error(
          `gatsby-plugin-advanced-pages: Unable to generate path for route "${to}": ${e.message}`,
        )
      }
      to = '/404.html'
    }
  }

  return <GatsbyLink to={to} {...(props as any)} />
}

Link.propTypes = {
  to: string.isRequired,
  params: object,
  scope: string,
}

export default Link
