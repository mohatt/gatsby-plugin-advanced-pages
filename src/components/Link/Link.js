import React from 'react'
import { Link as GatsbyLink } from 'gatsby'
import { string, object } from 'prop-types'
import { generatePath } from '../../api'

const Link = ({ to, params, scope, ...props }) => {
  let href
  if (to !== undefined) {
    try {
      href = generatePath(to, params, scope)
    } catch (e) {
      if (process.env.NODE_ENV !== `production`) {
        console.error(
          'Warning: Invalid route props supplied to `Link` component: ' + e.message
        )
      }
    }
  }

  return (
    <GatsbyLink to={href} {...props} />
  )
}

Link.propTypes = {
  to: string.isRequired,
  params: object,
  scope: string
}

export default Link
