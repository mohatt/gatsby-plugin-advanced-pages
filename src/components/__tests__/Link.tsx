// @vitest-environment jsdom
import React from 'react'
import ReactTest from '@testing-library/react'
import { Link, LinkProps } from '../Link'

describe('<Link />', () => {
  const consoleError = vi.spyOn(console, 'error').mockImplementation(() => {})

  beforeEach(() => {
    consoleError.mockClear()
  })

  const render = (props?: LinkProps<any>) => {
    return ReactTest.render(<Link {...props} />).container.firstChild
  }

  it('should render correctly', () => {
    expect(render({ to: 'home' })).toMatchSnapshot()
  })

  it('should render correctly with route params', () => {
    expect(
      render({
        to: 'blog.post',
        params: {
          post: 'hello-world',
        },
      }),
    ).toMatchSnapshot()
  })

  it('should render correctly with route params and scope', () => {
    expect(
      render({
        to: 'blog.tag',
        scope: 'pagination',
        params: {
          tag: 'news',
          page: 5,
        },
      }),
    ).toMatchSnapshot()
  })

  it('should render correctly with with no route prop', () => {
    expect(render()).toMatchSnapshot()
  })

  it('should log console error when called with invalid props', () => {
    render({ to: 'blog.post' })
    expect(consoleError).toHaveBeenCalledTimes(1)
    consoleError.mockClear()
    render({
      to: 'blog.post',
      params: {
        invalid: 'whatever',
      },
    })
    expect(consoleError).toHaveBeenCalledTimes(1)
  })

  it('should render correctly when called with a non-existent route', () => {
    expect(render({ to: 'invalid-route' })).toMatchSnapshot()
    expect(render({ to: 'some/path' })).toMatchSnapshot()
  })
})
