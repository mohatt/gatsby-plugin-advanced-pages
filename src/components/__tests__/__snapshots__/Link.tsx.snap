// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Link /> > should render correctly 1`] = `
<a
  href="/site/"
/>
`;

exports[`<Link /> > should render correctly when called with a non-existent route 1`] = `
<a
  href="/site/invalid-route"
/>
`;

exports[`<Link /> > should render correctly when called with a non-existent route 2`] = `
<a
  href="/site/some/path"
/>
`;

exports[`<Link /> > should render correctly with route params 1`] = `
<a
  href="/site/blog/post/hello-world"
/>
`;

exports[`<Link /> > should render correctly with route params and scope 1`] = `
<a
  href="/site/blog/tag/news/page/5"
/>
`;

exports[`<Link /> > should render correctly with with no route prop 1`] = `
<a
  href="?!"
/>
`;
